diff --git a/node_modules/@angular/build/src/tools/angular/compilation/angular-compilation.js b/node_modules/@angular/build/src/tools/angular/compilation/angular-compilation.js
index 53168ea..f07c80a 100755
--- a/node_modules/@angular/build/src/tools/angular/compilation/angular-compilation.js
+++ b/node_modules/@angular/build/src/tools/angular/compilation/angular-compilation.js
@@ -78,8 +78,6 @@ class AngularCompilation {
             allowEmptyCodegenFiles: false,
             annotationsAs: 'decorators',
             enableResourceInlining: false,
-            supportTestBed: false,
-            supportJitMode: false,
             // Disable removing of comments as TS is quite aggressive with these and can
             // remove important annotations, such as /* @__PURE__ */ and comments like /* vite-ignore */.
             removeComments: false,
diff --git a/node_modules/@angular/build/src/tools/esbuild/angular/compiler-plugin.js b/node_modules/@angular/build/src/tools/esbuild/angular/compiler-plugin.js
index ee68408..ac15cbf 100755
--- a/node_modules/@angular/build/src/tools/esbuild/angular/compiler-plugin.js
+++ b/node_modules/@angular/build/src/tools/esbuild/angular/compiler-plugin.js
@@ -90,7 +90,7 @@ function createCompilerPlugin(pluginOptions, compilationOrFactory, stylesheetBun
                 sourcemap: !!pluginOptions.sourcemap,
                 thirdPartySourcemaps: pluginOptions.thirdPartySourcemaps,
                 advancedOptimizations: pluginOptions.advancedOptimizations,
-                jit: pluginOptions.jit || pluginOptions.includeTestMetadata,
+                jit: true, //pluginOptions.jit || pluginOptions.includeTestMetadata,
             }, environment_options_1.maxWorkers, cacheStore?.createCache('jstransformer'));
             // Setup defines based on the values used by the Angular compiler-cli
             build.initialOptions.define ??= {};
@@ -448,12 +448,14 @@ function createCompilerPlugin(pluginOptions, compilationOrFactory, stylesheetBun
             async function hasSideEffects(path) {
                 if (!pluginOptions.advancedOptimizations) {
                     return undefined;
+                } else {
+                    return true;
                 }
-                const { sideEffects } = await build.resolve(path, {
+                /*const { sideEffects } = await build.resolve(path, {
                     kind: 'import-statement',
                     resolveDir: build.initialOptions.absWorkingDir ?? '',
                 });
-                return sideEffects;
+                return sideEffects;*/
             }
         },
     };
@@ -554,9 +556,7 @@ function createCompilerOptionsTransformer(setupWarnings, pluginOptions, preserve
             sourceRoot: undefined,
             preserveSymlinks,
             externalRuntimeStyles: pluginOptions.externalRuntimeStyles,
-            _enableHmr: !!pluginOptions.templateUpdates,
-            supportTestBed: !!pluginOptions.includeTestMetadata,
-            supportJitMode: !!pluginOptions.includeTestMetadata,
+            _enableHmr: !!pluginOptions.templateUpdates
         };
     };
 }
