diff --git a/node_modules/@iplab/ngx-color-picker/fesm2022/iplab-ngx-color-picker.mjs b/node_modules/@iplab/ngx-color-picker/fesm2022/iplab-ngx-color-picker.mjs
index a372799..f64a6f8 100644
--- a/node_modules/@iplab/ngx-color-picker/fesm2022/iplab-ngx-color-picker.mjs
+++ b/node_modules/@iplab/ngx-color-picker/fesm2022/iplab-ngx-color-picker.mjs
@@ -516,7 +516,7 @@ class Color {
         const s = (color.saturation / 100) * (l <= 1 ? l : 2 - l);
         const value = (l + s) / 2;
         const saturation = (2 * s) / (l + s) || 0;
-        return new Hsva(hue, saturation, value, color.alpha);
+        return new Hsva(hue, saturation * 100, value * 100, color.alpha);
     }
     rgbaToHsva(color) {
         const red = color.red / 255;
