diff --git a/node_modules/@mat-datetimepicker/core/esm2022/datetimepicker/clock.mjs b/node_modules/@mat-datetimepicker/core/esm2022/datetimepicker/clock.mjs
index 7ecfae7..08363d3 100644
--- a/node_modules/@mat-datetimepicker/core/esm2022/datetimepicker/clock.mjs
+++ b/node_modules/@mat-datetimepicker/core/esm2022/datetimepicker/clock.mjs
@@ -259,9 +259,9 @@ export class MatDatetimepickerClockComponent {
                 value = 0;
             }
             // Don't close the minutes view if an invalid minute is clicked.
-            if (!this._minutes.find((m) => m?.['value'] === value)?.['enabled']) {
-                return;
-            }
+            // if (!this._minutes.find((m) => m?.['value'] === value)?.['enabled']) {
+            //    return;
+            // }
             date = this._adapter.createDatetime(this._adapter.getYear(this.activeDate), this._adapter.getMonth(this.activeDate), this._adapter.getDate(this.activeDate), this._adapter.getHour(this.activeDate), value);
         }
         this._timeChanged = true;
diff --git a/node_modules/@mat-datetimepicker/core/fesm2022/mat-datetimepicker-core.mjs b/node_modules/@mat-datetimepicker/core/fesm2022/mat-datetimepicker-core.mjs
index 00f4a52..df688e3 100644
--- a/node_modules/@mat-datetimepicker/core/fesm2022/mat-datetimepicker-core.mjs
+++ b/node_modules/@mat-datetimepicker/core/fesm2022/mat-datetimepicker-core.mjs
@@ -946,9 +946,9 @@ class MatDatetimepickerClockComponent {
                 value = 0;
             }
             // Don't close the minutes view if an invalid minute is clicked.
-            if (!this._minutes.find((m) => m?.['value'] === value)?.['enabled']) {
-                return;
-            }
+            // if (!this._minutes.find((m) => m?.['value'] === value)?.['enabled']) {
+            //    return;
+            // }
             date = this._adapter.createDatetime(this._adapter.getYear(this.activeDate), this._adapter.getMonth(this.activeDate), this._adapter.getDate(this.activeDate), this._adapter.getHour(this.activeDate), value);
         }
         this._timeChanged = true;
