// recompose/index.js
'use strict';

exports.__esModule = true;
exports.setObservableConfig = exports.createEventHandler = exports.mapPropsStream = exports.componentFromStream = exports.hoistStatics = exports.nest = exports.componentFromProp = exports.createSink = exports.createEagerFactory = exports.createEagerElement = exports.isClassComponent = exports.shallowEqual = exports.wrapDisplayName = exports.getDisplayName = exports.compose = exports.setDisplayName = exports.setPropTypes = exports.setStatic = exports.toClass = exports.lifecycle = exports.getContext = exports.withContext = exports.onlyUpdateForPropTypes = exports.onlyUpdateForKeys = exports.pure = exports.shouldUpdate = exports.renderNothing = exports.renderComponent = exports.branch = exports.withReducer = exports.withState = exports.flattenProp = exports.renameProps = exports.renameProp = exports.defaultProps = exports.withHandlers = exports.withPropsOnChange = exports.withProps = exports.mapProps = undefined;

var _mapProps2 = require('./mapProps');

var _mapProps3 = _interopRequireDefault(_mapProps2);

var _withProps2 = require('./withProps');

var _withProps3 = _interopRequireDefault(_withProps2);

var _withPropsOnChange2 = require('./withPropsOnChange');

var _withPropsOnChange3 = _interopRequireDefault(_withPropsOnChange2);

var _withHandlers2 = require('./withHandlers');

var _withHandlers3 = _interopRequireDefault(_withHandlers2);

var _defaultProps2 = require('./defaultProps');

var _defaultProps3 = _interopRequireDefault(_defaultProps2);

var _renameProp2 = require('./renameProp');

var _renameProp3 = _interopRequireDefault(_renameProp2);

var _renameProps2 = require('./renameProps');

var _renameProps3 = _interopRequireDefault(_renameProps2);

var _flattenProp2 = require('./flattenProp');

var _flattenProp3 = _interopRequireDefault(_flattenProp2);

var _withState2 = require('./withState');

var _withState3 = _interopRequireDefault(_withState2);

var _withReducer2 = require('./withReducer');

var _withReducer3 = _interopRequireDefault(_withReducer2);

var _branch2 = require('./branch');

var _branch3 = _interopRequireDefault(_branch2);

var _renderComponent2 = require('./renderComponent');

var _renderComponent3 = _interopRequireDefault(_renderComponent2);

var _renderNothing2 = require('./renderNothing');

var _renderNothing3 = _interopRequireDefault(_renderNothing2);

var _shouldUpdate2 = require('./shouldUpdate');

var _shouldUpdate3 = _interopRequireDefault(_shouldUpdate2);

var _pure2 = require('./pure');

var _pure3 = _interopRequireDefault(_pure2);

var _onlyUpdateForKeys2 = require('./onlyUpdateForKeys');

var _onlyUpdateForKeys3 = _interopRequireDefault(_onlyUpdateForKeys2);

var _onlyUpdateForPropTypes2 = require('./onlyUpdateForPropTypes');

var _onlyUpdateForPropTypes3 = _interopRequireDefault(_onlyUpdateForPropTypes2);

var _withContext2 = require('./withContext');

var _withContext3 = _interopRequireDefault(_withContext2);

var _getContext2 = require('./getContext');

var _getContext3 = _interopRequireDefault(_getContext2);

var _lifecycle2 = require('./lifecycle');

var _lifecycle3 = _interopRequireDefault(_lifecycle2);

var _toClass2 = require('./toClass');

var _toClass3 = _interopRequireDefault(_toClass2);

var _setStatic2 = require('./setStatic');

var _setStatic3 = _interopRequireDefault(_setStatic2);

var _setPropTypes2 = require('./setPropTypes');

var _setPropTypes3 = _interopRequireDefault(_setPropTypes2);

var _setDisplayName2 = require('./setDisplayName');

var _setDisplayName3 = _interopRequireDefault(_setDisplayName2);

var _compose2 = require('./compose');

var _compose3 = _interopRequireDefault(_compose2);

var _getDisplayName2 = require('./getDisplayName');

var _getDisplayName3 = _interopRequireDefault(_getDisplayName2);

var _wrapDisplayName2 = require('./wrapDisplayName');

var _wrapDisplayName3 = _interopRequireDefault(_wrapDisplayName2);

var _shallowEqual2 = require('./shallowEqual');

var _shallowEqual3 = _interopRequireDefault(_shallowEqual2);

var _isClassComponent2 = require('./isClassComponent');

var _isClassComponent3 = _interopRequireDefault(_isClassComponent2);

var _createEagerElement2 = require('./createEagerElement');

var _createEagerElement3 = _interopRequireDefault(_createEagerElement2);

var _createEagerFactory2 = require('./createEagerFactory');

var _createEagerFactory3 = _interopRequireDefault(_createEagerFactory2);

var _createSink2 = require('./createSink');

var _createSink3 = _interopRequireDefault(_createSink2);

var _componentFromProp2 = require('./componentFromProp');

var _componentFromProp3 = _interopRequireDefault(_componentFromProp2);

var _nest2 = require('./nest');

var _nest3 = _interopRequireDefault(_nest2);

var _hoistStatics2 = require('./hoistStatics');

var _hoistStatics3 = _interopRequireDefault(_hoistStatics2);

var _componentFromStream2 = require('./componentFromStream');

var _componentFromStream3 = _interopRequireDefault(_componentFromStream2);

var _mapPropsStream2 = require('./mapPropsStream');

var _mapPropsStream3 = _interopRequireDefault(_mapPropsStream2);

var _createEventHandler2 = require('./createEventHandler');

var _createEventHandler3 = _interopRequireDefault(_createEventHandler2);

var _setObservableConfig2 = require('./setObservableConfig');

var _setObservableConfig3 = _interopRequireDefault(_setObservableConfig2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.mapProps = _mapProps3.default; // Higher-order component helpers

exports.withProps = _withProps3.default;
exports.withPropsOnChange = _withPropsOnChange3.default;
exports.withHandlers = _withHandlers3.default;
exports.defaultProps = _defaultProps3.default;
exports.renameProp = _renameProp3.default;
exports.renameProps = _renameProps3.default;
exports.flattenProp = _flattenProp3.default;
exports.withState = _withState3.default;
exports.withReducer = _withReducer3.default;
exports.branch = _branch3.default;
exports.renderComponent = _renderComponent3.default;
exports.renderNothing = _renderNothing3.default;
exports.shouldUpdate = _shouldUpdate3.default;
exports.pure = _pure3.default;
exports.onlyUpdateForKeys = _onlyUpdateForKeys3.default;
exports.onlyUpdateForPropTypes = _onlyUpdateForPropTypes3.default;
exports.withContext = _withContext3.default;
exports.getContext = _getContext3.default;
exports.lifecycle = _lifecycle3.default;
exports.toClass = _toClass3.default;

// Static property helpers

exports.setStatic = _setStatic3.default;
exports.setPropTypes = _setPropTypes3.default;
exports.setDisplayName = _setDisplayName3.default;

// Composition function

exports.compose = _compose3.default;

// Other utils

exports.getDisplayName = _getDisplayName3.default;
exports.wrapDisplayName = _wrapDisplayName3.default;
exports.shallowEqual = _shallowEqual3.default;
exports.isClassComponent = _isClassComponent3.default;
exports.createEagerElement = _createEagerElement3.default;
exports.createEagerFactory = _createEagerFactory3.default;
exports.createSink = _createSink3.default;
exports.componentFromProp = _componentFromProp3.default;
exports.nest = _nest3.default;
exports.hoistStatics = _hoistStatics3.default;

// Observable helpers

exports.componentFromStream = _componentFromStream3.default;
exports.mapPropsStream = _mapPropsStream3.default;
exports.createEventHandler = _createEventHandler3.default;
exports.setObservableConfig = _setObservableConfig3.default;
