diff --git a/src/test/java/org/thingsboard/server/msa/edge/CalculatedFieldTest.java b/src/test/java/org/thingsboard/server/msa/edge/CalculatedFieldTest.java
new file mode 100644
index 0000000..ed36c43
--- /dev/null
+++ b/src/test/java/org/thingsboard/server/msa/edge/CalculatedFieldTest.java
@@ -0,0 +1,147 @@
+/**
+ * Copyright © 2016-2025 The Thingsboard Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.msa.edge;
+
+import org.junit.Test;
+import org.testcontainers.shaded.org.awaitility.Awaitility;
+import org.thingsboard.server.common.data.Device;
+import org.thingsboard.server.common.data.EntityType;
+import org.thingsboard.server.common.data.cf.CalculatedField;
+import org.thingsboard.server.common.data.cf.CalculatedFieldType;
+import org.thingsboard.server.common.data.cf.configuration.Argument;
+import org.thingsboard.server.common.data.cf.configuration.ArgumentType;
+import org.thingsboard.server.common.data.cf.configuration.Output;
+import org.thingsboard.server.common.data.cf.configuration.OutputType;
+import org.thingsboard.server.common.data.cf.configuration.ReferencedEntityKey;
+import org.thingsboard.server.common.data.cf.configuration.SimpleCalculatedFieldConfiguration;
+import org.thingsboard.server.common.data.debug.DebugSettings;
+import org.thingsboard.server.common.data.group.EntityGroup;
+import org.thingsboard.server.common.data.id.DeviceId;
+import org.thingsboard.server.common.data.id.EntityId;
+import org.thingsboard.server.common.data.page.PageLink;
+import org.thingsboard.server.gen.edge.v1.EdgeVersion;
+import org.thingsboard.server.msa.AbstractContainerTest;
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class CalculatedFieldTest extends AbstractContainerTest {
+
+    @Test
+    public void testSendCalculatedFieldToEdge() {
+        performTestOnEachEdge(this::_testSendCalculatedFieldToEdge, EdgeVersion.V_4_1_0);
+    }
+
+    private void _testSendCalculatedFieldToEdge() {
+        EntityGroup savedDeviceEntityGroup1 = createEntityGroup(EntityType.DEVICE);
+        Device savedDevice = saveDeviceAndAssignEntityGroupToEdge("Default CalculatedFiled", savedDeviceEntityGroup1);
+        DeviceId savedDeviceId = savedDevice.getId();
+
+        String calculatedFieldName = "CalculatedField";
+        SimpleCalculatedFieldConfiguration config = new SimpleCalculatedFieldConfiguration();
+        CalculatedField calculatedField = createSimpleCalculatedField(savedDevice.getId(), config, calculatedFieldName);
+
+        // create calculatedField on cloud
+        CalculatedField savedCalculatedField = cloudRestClient.saveCalculatedField(calculatedField);
+        Awaitility.await()
+                .pollInterval(500, TimeUnit.MILLISECONDS)
+                .atMost(30, TimeUnit.SECONDS)
+                .until(() -> edgeRestClient.getCalculatedFieldsByEntityId(savedDeviceId, new PageLink(1)).getTotalElements() == 1);
+
+        // update calculatedField on cloud
+        String updateName = calculatedFieldName + " Updated";
+        savedCalculatedField.setName(updateName);
+        cloudRestClient.saveCalculatedField(savedCalculatedField);
+        Awaitility.await()
+                .pollInterval(500, TimeUnit.MILLISECONDS)
+                .atMost(30, TimeUnit.SECONDS)
+                .until(() -> edgeRestClient.getCalculatedFieldsByEntityId(savedDeviceId, new PageLink(1)).getData().get(0).getName().equals(updateName));
+
+        // delete calculatedField on cloud
+        cloudRestClient.deleteCalculatedField(savedCalculatedField.getId());
+        Awaitility.await()
+                .pollInterval(500, TimeUnit.MILLISECONDS)
+                .atMost(30, TimeUnit.SECONDS)
+                .until(() -> edgeRestClient.getCalculatedFieldsByEntityId(savedDeviceId, new PageLink(1)).getData().isEmpty());
+    }
+
+    @Test
+    public void testSendCalculatedFieldToCloud() {
+        performTestOnEachEdge(this::_testSendCalculatedFieldToCloud, EdgeVersion.V_4_1_0);
+    }
+
+    private void _testSendCalculatedFieldToCloud() {
+        EntityGroup savedDeviceEntityGroup1 = createEntityGroup(EntityType.DEVICE);
+        Device savedDevice = saveDeviceAndAssignEntityGroupToEdge("Default CalculatedFiled", savedDeviceEntityGroup1);
+        DeviceId savedDeviceId = savedDevice.getId();
+
+        String calculatedFieldName = "CalculatedField";
+        SimpleCalculatedFieldConfiguration config = new SimpleCalculatedFieldConfiguration();
+        CalculatedField calculatedField = createSimpleCalculatedField(savedDevice.getId(), config, calculatedFieldName);
+
+
+        // create calculatedField on edge
+        CalculatedField savedCalculatedField = edgeRestClient.saveCalculatedField(calculatedField);
+        Awaitility.await()
+                .pollInterval(500, TimeUnit.MILLISECONDS)
+                .atMost(30, TimeUnit.SECONDS)
+                .until(() -> cloudRestClient.getCalculatedFieldsByEntityId(savedDeviceId, new PageLink(1)).getTotalElements() == 1);
+
+        // update calculatedField on cloud
+        String updateName = calculatedFieldName + " Updated";
+        savedCalculatedField.setName(updateName);
+        edgeRestClient.saveCalculatedField(savedCalculatedField);
+        Awaitility.await()
+                .pollInterval(500, TimeUnit.MILLISECONDS)
+                .atMost(30, TimeUnit.SECONDS)
+                .until(() -> cloudRestClient.getCalculatedFieldsByEntityId(savedDeviceId, new PageLink(1)).getData().get(0).getName().equals(updateName));
+
+        // delete calculatedField on edge
+        edgeRestClient.deleteCalculatedField(savedCalculatedField.getId());
+        Awaitility.await()
+                .pollInterval(500, TimeUnit.MILLISECONDS)
+                .atMost(30, TimeUnit.SECONDS)
+                .until(() -> cloudRestClient.getCalculatedFieldsByEntityId(savedDeviceId, new PageLink(1)).getData().isEmpty());
+    }
+
+    private CalculatedField createSimpleCalculatedField(EntityId entityId, SimpleCalculatedFieldConfiguration config, String name) {
+        CalculatedField calculatedField = new CalculatedField();
+        calculatedField.setEntityId(entityId);
+        calculatedField.setTenantId(edge.getTenantId());
+        calculatedField.setType(CalculatedFieldType.SIMPLE);
+        calculatedField.setName(name);
+        calculatedField.setDebugSettings(DebugSettings.all());
+
+        Argument argument = new Argument();
+        ReferencedEntityKey refEntityKey = new ReferencedEntityKey("temperature", ArgumentType.TS_LATEST, null);
+        argument.setRefEntityKey(refEntityKey);
+        argument.setDefaultValue("12"); // not used because real telemetry value in db is present
+        config.setArguments(Map.of("T", argument));
+
+        config.setExpression("(T * 9/5) + 32");
+
+        Output output = new Output();
+        output.setName("fahrenheitTemp");
+        output.setType(OutputType.TIME_SERIES);
+        output.setDecimalsByDefault(2);
+        config.setOutput(output);
+
+        calculatedField.setConfiguration(config);
+
+        return calculatedField;
+    }
+
+}
